### PR TITLE
Rate limiting

### PR DESCRIPTION
**Pull request by talyguryn**

### CODE REVIEW

File: capella/src/controller/Form.php [hunk 1]
```diff
@@ -14,6 +16,20 @@ class Form
     public function __construct()
     {
         if ($_SERVER['REQUEST_METHOD'] === 'POST') {
+            $ip = Methods::getRequestSourceIp();
+            $key = 'uploadsBy_' . $ip;
```
Comment (neSpecc) [inline capella/src/controller/Form.php]: move key composing to separated method

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: capella/src/controller/Form.php (hunk 1 above)
Comment (talyguryn) [reply capella/src/controller/Form.php]: yea, of course it will be done

File: capella/src/RateLimiter.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
+<?php
+
```
Comment (khaydarov) [inline capella/src/RateLimiter.php:2]:
```suggestion
```

File: capella/src/api/Response.php [hunk 3]
```diff
@@ -14,7 +14,7 @@ class Response
      */
     public static function json($responseData)
     {
-        echo json_encode($responseData);
+        die(json_encode($responseData));
```
Comment (khaydarov) [inline capella/src/api/Response.php]: это ты так возвращаешь HTTP-ответ?)

File: capella/src/cache/Cache.php [hunk 4]
```diff
@@ -116,6 +126,23 @@ public function delete($key)
...
 
+    /**
+     * Increment item's value
+     *
+     * @param string $key   - cache key
+     * @param int    $value - increment the item by value
+     */
+    public function increment($key, $value = 1)
+    {
+        if (is_null($this->memcacheObj)) {
```
Comment (gohabereg) [inline capella/src/cache/Cache.php]: Doesn't `isAlive` method do the same?

File: capella/src/controller/Form.php [hunk 5]
```diff
@@ -115,4 +121,24 @@ protected function returnImageData($imageData)
...
+            $ip = Methods::getRequestSourceIp();
+            $key = 'RATELIMITER_CLIENT_' . $ip;
+
+            $requestAllowed = RateLimiter::instance()->check($key);
+
+            if (!$requestAllowed) {
+                HTTP\Response::TooManyRequests();
+
+                API\Response::error([
+                    'message' => 'Sorry, you cannot upload more than 3 images per minute.'
```
Comment (gohabereg) [inline capella/src/controller/Form.php]: I guess you can use constants from ENV here

File: capella/src/http/Response.php [hunk 6]
```diff
@@ -38,6 +38,12 @@ public static function MethodNotAllowed()
         self::response(405, 'Method Not Allowed');
     }
 
+    public static function TooManyRequests($retryAfter = 60)
```
Comment (gohabereg) [inline capella/src/http/Response.php]: Here also constant can be used

**Review by gohabereg — CHANGES_REQUESTED** [review]

**Review by gohabereg — APPROVED** [review]

File: capella/src/controller/Form.php [hunk 7]
```diff
@@ -14,6 +16,10 @@ class Form
     public function __construct()
     {
         if ($_SERVER['REQUEST_METHOD'] === 'POST') {
+            /** Middleware to reduce image download intensity */
```
Comment (neSpecc) [inline capella/src/controller/Form.php]:
```suggestion
            /** Middleware to reduce image upload intensity */
```

**Review by neSpecc — APPROVED** [review]

**Pull request merged**